Derived templates

kallewoof opened the pull request:
This PR adds a simple hash / sub-string based method for picking context and instruct templates based on the chat template, when provided by the back end.

It can be confusing to have to manually switch when you switch to a different model, so having an automated way can be helpful.

The koboldcpp backend variant relies on https://github.com/LostRuins/koboldcpp/pull/1222 (merged, hopefully in 1.79 release).
The llama.cpp backend variant works as is.

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).
- [x] Add UI/option to enable or disable the auto-toggling.
- [x] Optional: add more hash-to-template derivations.

## Future work (post-merge):

* This could be expanded to **warn** the user when the template and the derived template are not compatible (as opposed to being all-or-nothing auto-swap like it is now).
* Since the chat template is provided as is, it could also be made more sophisticated, e.g. a parser that compares the current preset with the provided one and determines if they are compatible.
* Allow users to map hash (model type) to whatever preset they prefer. Ideally, remember the decision a user makes when they pick a model given a chat template, and use that model for that model / chat template in the future. Users may want to use custom presets for a specific model which shares a chat template with others so not sure what the ideal approach would be here.

kallewoof committed:
feature: derived templates

This PR adds a simple hash based method for picking context and instruct templates based on the chat template, when provided by the back end.

Cohee1207 commented on public/scripts/chat-cemplates.js:
```diff
@@ -0,0 +1,76 @@
+// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest
+async function digestMessage(message) {
+    const msgUint8 = new TextEncoder().encode(message); // encode as (utf-8) Uint8Array
+    const hashBuffer = await window.crypto.subtle.digest('SHA-256', msgUint8); // hash the message
```
[window.crypto.subtle](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/subtle#browser_compatibility) is only available in secure contexts (i.e. localhost or HTTPS).
But you can import any library from npm using Webpack, for example [js-sha256](https://www.npmjs.com/package/js-sha256)

kallewoof replied:
Thanks for pointing that out. I put the hash into the response from the backend to avoid having to bundle anything. It looks reasonable to me, but if you prefer Webpack style I can do that.

Cohee1207 replied:
Doing it on a backend is fine too. crypto module is always available in Node.

kallewoof replied:
Right, one less dependency. Switched to using that.

kallewoof committed:
template derivation: move hash part to backend

kallewoof committed:
switch to crypto lib

kallewoof committed:
fix error console.log message

kallewoof committed:
template derivation: add support for llama.cpp server backend

kallewoof commented:
So, it turns out that llama.cpp already supports fetching the chat template, so adding llama.cpp server support was trivial.

There's a tiny gotcha, where despite returning JSON data, the chat_template string from llama.cpp ends with a NUL terminator. Clearly a bug, so once that is fixed, the corresponding code can be discarded.

kallewoof committed:
UI: add UI to enable/disable auto-derived templates

kallewoof committed:
update endpoint to reflect koboldcpp update

kallewoof commented:
The koboldcpp pull request was updated to use the same endpoint as llama.cpp (/props), which means both can be treated the same way, in terms of chat template fetching.

I also added UI to enable/disable the derived template feature. It defaults to off, for now.

kallewoof committed:
trim chat template before hashing

kallewoof commented:
A note on the last commit: there is [a bug in llama.cpp](https://github.com/ggerganov/llama.cpp/pull/10419) which results in the chat template string  being null-terminated one character too early. This has two consequences: it means the last character (which happens to be a `\n` in every case) is trimmed out, and it means a `\u0000` is included in the resulting JSON string, in some cases. The latter is handled in feb1b91 already (mentioned in https://github.com/SillyTavern/SillyTavern/pull/3090#issuecomment-2485903551 above), and the former becomes a compatibility thing. I noted at the top in chat-templates.js that
```js
// the hash can be obtained from command line e.g. via: MODEL=path_to_model; python -c "import json, hashlib, sys; print(hashlib.sha256(json.load(open('"$MODEL"/tokenizer_config.json'))['chat_template'].strip().encode()).hexdigest())"
// note that chat templates must be trimmed to match the llama.cpp metadata value
```
which was the case (due to this bug), but it is not the case anymore.

Approaches:

1. Keep the string trimming. Pros: easier to 'get right', avoids issues with llama.cpp/koboldcpp before/after the patch above -- all versions will work fine. Cons: annoying indefinite remnants.
2. Re-calculate all hashes without trimming, using the correct chat templates. Pros: makes things simpler. Cons: need to juggle broken chat template responses (for awhile), but probably not indefinitely.

kallewoof committed:
rename /chat_template to /props, and switch to non-trimmed hashes

kallewoof commented:
I went with approach 2. It seems like the cleaner approach, even though string trimming is fairly common. The llama.cpp NULL value is now replaced with a newline (which it originally replaced), which aligns it with the koboldcpp endpoint.

I also renamed the Silly Tavern back end endpoint to /props, since that's what it is forwarding right now (not just chat template, at least for llama.cpp version).

kallewoof committed:
added substr derivations with one example

Cohee1207 commented:
How ready is this for a final review/test?

kallewoof commented:
> How ready is this for a final review/test?

It should be good to go!

Cohee1207 committed:
Merge branch 'staging' into 202411-auto-templates

Cohee1207 committed:
Update documentation links

Cohee1207 commented on public/index.html:
```diff
@@ -3295,9 +3298,12 @@ <h3 class="margin0 flex1" data-i18n="Advanced Formatting">
+                                <div class="flex-container">
+                                    <label for="context_derived" class="checkbox_label flex1" title="Derive from Model Metadata, if possible." data-i18n="[title]context_derived">
+                                        <input id="context_derived" type="checkbox" style="display:none;" />
+                                        <small><i class="fa-solid fa-tree menu_button margin0"></i></small>
```
I'm not sure that tree is the most obvious choice (and it looks like a bell from afar). I will try 'bolt' as it seems to have some associations with automation and scripting.

kallewoof replied:
Yeah, I wasn't sure either. Bolt sounds good to me!

Cohee1207 committed:
Change icon for derivation

Cohee1207 committed:
Check for response ok

Cohee1207 committed:
Add 'node:' prefix to import

Cohee1207 committed:
Exclude 'derived' from instruct templates

Cohee1207 committed:
Fix typo in file name

Cohee1207 committed:
Don't log when no endpoint available. Remove pointless header

Cohee1207 committed:
Add template hash for gemma-2-2b

Cohee1207 committed:
Validate instruct/context existance

Cohee1207 committed:
Remove custom header

Cohee1207 commented on src/endpoints/backends/text-completions.js:
```diff
@@ -218,6 +219,9 @@ router.post('/status', jsonParser, async function (request, response) {
             } catch (error) {
                 console.error(`Failed to get TabbyAPI model info: ${error}`);
             }
+        } else if (apiType == TEXTGEN_TYPES.KOBOLDCPP || apiType == TEXTGEN_TYPES.LLAMACPP) {
```
This custom header part was not needed, so I removed it.
The pattern was probably copied from tokenization-supported which only exists because of LM Studio-specific hacks.

kallewoof replied:
Got it. Works for me!

Cohee1207 reviewed (approved):
Tested with the latest pull of llama.cpp. Works, thanks!

Now we need someone to keep the mapping in sync with the latest models in town, lol.

kallewoof commented:
> Tested with the latest pull of llama.cpp. Works, thanks!

Thanks for testing/merging.

> Now we need someone to keep the mapping in sync with the latest models in town, lol.

Right. Seems like not a big deal esp since it console.logs the unknown hashes, but if it turns into a huge issue, it may be time to consider a more sophisticated approach. Doubt it though.
